worker/uniter/container/workload.go

```diff
@@ -0,0 +1,221 @@
+
+// WorkloadEvents is an interface providing a means of storing and retrieving
+// arguments for running workload hooks.
+type WorkloadEvents interface {
+	// AddWorkloadEvent adds the given command arguments and response function
+	// and returns a unique identifier.
+	AddWorkloadEvent(evt WorkloadEvent, cb WorkloadEventCallback) string
+
+	// GetWorkloadEvent returns the command arguments and response function
+	// with the specified ID, as registered in AddWorkloadEvent.
+	GetWorkloadEvent(id string) (WorkloadEvent, WorkloadEventCallback, error)
+
+	// RemoveWorkloadEvent removes the command arguments and response function
+	// associated with the specified ID.
+	RemoveWorkloadEvent(id string)
+
+	// Events returns all the currently queued events pending processing.
+	// Useful for debugging/testing.
+	Events() []WorkloadEvent
+}
+
+type workloadEvents struct {
+	mu      sync.Mutex
+	nextID  int
+	pending map[string]workloadEventItem
+}
+
+type workloadEventItem struct {
+	WorkloadEvent
+	cb WorkloadEventCallback
+}
+
+func NewWorkloadEvents() WorkloadEvents {
+	return &workloadEvents{pending: make(map[string]workloadEventItem)}
+}
+
+func (c *workloadEvents) AddWorkloadEvent(evt WorkloadEvent, cb WorkloadEventCallback) string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	id := strconv.Itoa(c.nextID)
+	c.nextID++
+	c.pending[id] = workloadEventItem{evt, cb}
+	return id
+}
+
+func (c *workloadEvents) RemoveWorkloadEvent(id string) {
+	c.mu.Lock()
+	delete(c.pending, id)
+	c.mu.Unlock()
+}
+
+func (c *workloadEvents) GetWorkloadEvent(id string) (WorkloadEvent, WorkloadEventCallback, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	item, ok := c.pending[id]
+	if !ok {
+		return WorkloadEvent{}, nil, errors.NotFoundf("workload event %s", id)
+	}
+	return item.WorkloadEvent, item.cb, nil
+}
+
+func (c *workloadEvents) Events() []WorkloadEvent {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	events := []WorkloadEvent(nil)
+	for _, v := range c.pending {
+		events = append(events, v.WorkloadEvent)
+	}
+	return events
+}
+
+type workloadHookResolver struct {
+	logger         Logger
+	events         WorkloadEvents
+	eventCompleted func(id string)
+}
+
+// NewWorkloadHookResolver returns a new resolver with determines which workload related operation
+// should be run based on local and remote uniter states.
+func NewWorkloadHookResolver(logger Logger, events WorkloadEvents, eventCompleted func(string)) resolver.Resolver {
+	return &workloadHookResolver{
+		logger:         logger,
+		events:         events,
+		eventCompleted: eventCompleted,
+	}
+}
+
+// NextOp implements the resolver.Resolver interface.
+func (r *workloadHookResolver) NextOp(
+	localState resolver.LocalState,
+	remoteState remotestate.Snapshot,
+	opFactory operation.Factory,
+) (operation.Operation, error) {
+	noOp := func() (operation.Operation, error) {
+		if localState.Kind == operation.RunHook &&
+			localState.Hook != nil &&
+			localState.Hook.Kind == hooks.WorkloadReady {
```

Comment:
If a workload hook was inflight and we get interrupted, doesn't this mean the hook will end up not being run on a restart?

Reply:
Will revisit this in the next PR.